fw/websocket.c

```diff
@@ -150,10 +150,9 @@ tfw_ws_srv_new_steal_sk(TfwSrvConn *srv_conn)
 	 */
 	conn->sk->sk_user_data = conn;
 
-	/* Connection destructor does failover for server connections */
 	srv_conn->sk = NULL;
-	if (srv_conn->destructor)
-		srv_conn->destructor(srv_conn);
+	/* Connection destructor does failover for server connections */
+	tfw_connection_put((TfwConn *)srv_conn);
```

Comment:
We did not decrement the reference counter because the connection is still used: the function 'steals' the high level connection decriptor `TfwSrvCon`, i.e. it leave it alive, but replaces the underlying HTTP backend server connection with a websocket. I.e. technically we still should have non-zero reference counter here. I'd expect that with the change we might have double free issue or get something like https://github.com/tempesta-tech/tempesta/issues/1647 - trying to send data proxied from a client websocket to the dead server websocket connection

Reply:
Looks like the most safe way to do this steam is: replace `tfw_connection_put` with `tfw_connection_put_to_death`
this will set refcounter to `TFW_CONN_DEATHCNT + 1` and then call the destructor. If `tfw_connection_put` will be called from `tfw_sock_srv_connect_drop` before `tfw_sock_srv_connect_try` we will have `refcounter == TFW_CONN_DEATHCNT` that good for us. Also `TFW_CONN_DEATHCNT + 1` is safe for as while reconnecting, because refcounter will be setted to 1 when connection successfully established.

Reply:
No, I wrong. `tfw_sock_srv_connect_drop` will not be called, because we don't have socket at this time. `tfw_ws_srv_new_steal_sk` will be only one place where we call `tfw_connection_put` for stolen connection.

Reply:
We can put connection's counter because during reconnecting refcounter will be updated. Also `ss_close` immediately return since conn->sk = NULL so we newer call connection_drop again until we restore conn->sk after successful reconnect.